### ISSUE TITLE
Revising Control.Biegunka.Source.Git repertoire 

### ISSUE THREAD

**Issue by dmalikov** [biegunka/biegunka#66]
```haskell
-- | Do nothing except pulling @origin/master@ into @master@
defaultGit :: Git
defaultGit = Git
  { _actions = return ()
  , _remote  = "origin"
  , _branch  = "master"
  }
```

Currently, `updateGit` is reseting to `origin/master` instead of pulling it, or even rebasing (which is actually looks more convenient here).

There are not so many tests covering this behaviour, so it's hard to understand what type of functionality is supposed to be correct and predictable.

Let me describe my vision of how `git{,',_}` functions should behave (`Actions` part seems to be obvious, thus I'll skip it). Feel free to correct me if you find anything wrong.

---
`git url path & remote .~ "remoteName" . branch "branchName"`

1) `path` does not exist

Expected behaviour - clone `branchName` of a `url` setting remote name to `remoteName`.

Actual behavior - `remoteName` is ignored. Failure!

2) `path` exists, repository have clean state, there are no commits in `branchName` ahead of `remoteName/branchName`

Expected behaviour - update `branchName` to `remoteName/branchName` (also, `remoteName` should be added).

Actual behavior - current implementation of `updateGit` fails if `remoteName` doesn't exist. Failure!

3) `path` exists, repository have clean state, there are some commits in `branchName` ahead of `remoteName/branchName`

Expected behaviour - rebase `branchName` onto `remoteName/branchName`.

Actual behavoir - `branchName` is reset to `remoteName/branchName`. Failure!

4) `path` exists, repository have dirty state

Expected behavior - error.

Actual behavior - error. Success!

---

What do you think about it?

**Comment by supki** [comment]
Can we simplify this by dropping the customization of the remote name and always using `origin` instead? What would be a use case for that option?

**Comment by supki** [comment]
I would prefer the update to fail in the case 3, actually. Perhaps that could be an option?

**Comment by supki** [comment]
Other than that LGTM. The current state is the result of evolution and is not particularly great. Any documented and tested update algorithm is probably an improvement.

**Comment by dmalikov** [comment]
> Can we simplify this by dropping the customization of the remote name and always using origin instead? What would be a use case for that option?

It's hard to imagine any case, when you need to update non-`origin` remote only, even if you have many of these remotes. So yeah, it seems to be pretty useless.

> I would prefer the update to fail in the case 3, actually. Perhaps that could be an option?

Could you please explain a little bit what is the point of failing there?
My desire of using `rebase` instead of `reset` is caused by necessity to be able to run biegunka's pointed at local dotfiles repository containing commits which are not pushed yet.

Actually, I though that for case like that is better to use `Source.directory` pointed at the `$pwd` where biegunka is called from (which is also remove the silly strictness to locate dotfiles repository in a same filepath across all environment). But I stuck on `IO ()` produced by `getCurrentDirectory` which is polluting all pure `Namespace`s layout.

> Any documented and tested update algorithm is probably an improvement.

I'm not quite a fan of documentation because it always appears almost outdated. It will be great to have a couple of tests on `Git` module, but I'm not quite sure what approach we should use here.

**Comment by supki** [comment]
> Could you please explain a little bit what is the point of failing there?

Sure thing. While automatic rebasing is fine for repositories I control (especially for e.g. purely dotfilesy repositories), I don't trust myself to make changes to others' work that withstand the test of time and don't break anything on updates silently (even when they apply cleanly). So I'd rather look at their changes before I rebase mine.

**Comment by supki** [comment]
As for documentation, it could be as simple as a set of specs (similar to `spec/typecheck`) as long as it's clear what's being tested.

**Comment by supki** [comment]
Also, could you make a separate issue for the problem with `getCurrentDirectory` you described? It isn't clear to me what it is (yet).

**Comment by dmalikov** [comment]
Since #67 is merged, let's define a scope of changes that need to be done:
- [ ] make `updateGit` check that current branch matches the one `biegunka` going to checkout
- [ ] make `updateGit` check that remote url matches the one `biegunka` going to fetch from
- [ ] add one happy-case integration test checking whole `biegunka`'s flow 

Anything else?

**Comment by supki** [comment]
It'd be great to merge the `git*` family of functions into a single function like `biegunka-svn` does.

**Comment by dmalikov** [comment]
Probably there is nothing left to be done in this task.

**Comment by supki** [comment]
Nice work! :+1:

**Comment by supki** [comment]
I think there's a problem. Once you have unpublished local commits, a bogus diff of the form `$unpublished_changes_ahead_of_remote → $remote` will be displayed every run.

**Comment by dmalikov** [comment]
Yep, this is not really a helpful diff status message thrown from `C.B.S.G.update` function. What would be an appropriate message in this case?

Also, if I understand the issue correctly, it's related to changes made in @4007acae

**Comment by supki** [comment]
The aforementioned problem has been fixed in 4ef5d67.
